### DISCUSSION TITLE
message timestamp and clock skew

### DISCUSSION THREAD

**Discussion by stereosteve** [nats-io/nats-server#3095] [Q&A] [answered]
Hello - 

Was curious to better understand the variability of Timestamp field in [MsgMetadata](https://pkg.go.dev/github.com/nats-io/nats.go#MsgMetadata).

Is that assigned using the leader's system clock?  Does NATS jetstream cluster attempt to detect clock skew?

**Comment by derekcollison** [comment, marked as answer]
Yes that is set by the leader's clock and we require clocks to be fairly synchronized with something liek ntp.
